src/main/scala/com/databricks/labs/overwatch/pipeline/DbsqlTransforms.scala

```diff
@@ -0,0 +1,130 @@
+package com.databricks.labs.overwatch.pipeline
+
+import com.databricks.labs.overwatch.pipeline.PipelineFunctions.fillForward
+import com.databricks.labs.overwatch.pipeline.TransformFunctions._
+import com.databricks.labs.overwatch.utils.SchemaTools.structFromJson
+import com.databricks.labs.overwatch.utils._
+import org.apache.log4j.{Level, Logger}
+import org.apache.spark.sql.expressions.{Window, WindowSpec}
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{Column, DataFrame}
+
+object DbsqlTransforms extends SparkSessionWrapper {
+
+  import spark.implicits._
+  private val logger: Logger = Logger.getLogger(this.getClass)
+  val responseSuccessFilter: Column = $"response.statusCode" === 200
+
+  /**
+   * BEGIN DBSQL generic functions
+   */
+
+  private val auditBaseCols: Array[Column] = Array(
+    'timestamp, 'date, 'organization_id, 'serviceName, 'actionName,
+    $"userIdentity.email".alias("userEmail"), 'requestId, 'response)
+
+  def deriveWarehouseId(): Column = {
+    when(('actionName === "createEndpoint" || 'actionName === "createWarehouse")
+      && responseSuccessFilter,
+      get_json_object($"response.result", "$.id"))
+      .otherwise('id)
+  }
+
+  def deriveWarehouseBase(auditRawDF: DataFrame): DataFrame = {
+    val warehouse_id_gen_w = Window.partitionBy('organization_id, 'warehouse_name)
+      .orderBy('timestamp).rowsBetween(Window.currentRow, 1000)
+    val warehouse_name_gen_w = Window.partitionBy('organization_id, 'warehouse_id)
+      .orderBy('timestamp).rowsBetween(Window.currentRow, 1000)
+
+    val warehouseSummaryCols = auditBaseCols ++ Array[Column](
+      deriveWarehouseId.alias("warehouse_id"),
+      'name.alias("warehouse_name"),
+      'cluster_size,
+      'min_num_clusters.cast("long"),
+      'max_num_clusters.cast("long"),
+      'auto_stop_mins.cast("long"),
+      'spot_instance_policy,
+      'enable_photon.cast("boolean"),
+      'channel,
+      'tags,
+      'enable_serverless_compute.cast("boolean"),
+      'warehouse_type
+    )
+
+    val warehouseRaw = auditRawDF
+      .filter('serviceName === "databrickssql")
+      .selectExpr("*", "requestParams.*").drop("requestParams", "Overwatch_RunID")
+      .select(warehouseSummaryCols: _*)
+      .withColumn("warehouse_id", PipelineFunctions.fillForward("warehouse_id", warehouse_id_gen_w))
+      .withColumn("warehouse_name", PipelineFunctions.fillForward("warehouse_name", warehouse_name_gen_w))
```

Comment:
no. First, warehouse_id should never be null; thus you don't want to try and derive id from name. Warehouse_name may be necessary to fillForward but only if there are nulls for some of the actionNames in scope. Also, if name is null; there are likely several other fields that are null as well and for those you'll need to fill forward but you shouldn't do this with a withColumn each time as that creates a very nasty spark plan. Each of these fillForward functions should be part of the selects. If any are still unfillable you can use the warehouseSnap as a secondary fill. See the `poolSelects` in SilverTransforms.scala as an example.

Reply:
done. Also passing full silver_spec_warehouse to fill forward appropriately.